data-science-onramp/data-cleaning/clean.py

```diff
@@ -123,69 +123,61 @@ def compute_end_udf(duration, start, end):
     '''Calculates end time from duration and start time if null.'''
     return compute_time(duration, start, end)[2]
 
-def upload_to_gcs(df):
-    '''Writes a dataframe to a GCP bucket.'''
-    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
-    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
-    print('Data successfully uploaded to ' + path)
 
-def main():
-    # Create a SparkSession under the name 'data_cleaning'. Viewable via the Spark UI
+if __name__ == '__main__':
+    # Create a SparkSession, viewable via the Spark UI
     spark = SparkSession.builder.appName('data_cleaning').getOrCreate()
 
-    # Whether to upload data to GCS bucket
-    upload = True
-    if '--test' in sys.argv:
-        upload = False
-        print('Results will not be uploaded to GCS')
-
-    # Check if table exists
+    # Load data into dataframe if table exists
     try:
         df = spark.read.format('bigquery').option('table', TABLE).load()
     except Py4JJavaError:
-        print(f'{TABLE} does not exist. ')
-        return
+        print(f'{TABLE} does not exist.')
+        raise 
 
-    # Single-parameter column transformations
+    # Single-parameter udfs
     udfs = {
-        'start_station_name': (station_name_udf, StringType()),
-        'end_station_name': (station_name_udf, StringType()),
-        'tripduration': (trip_duration_udf, IntegerType()),
-        'usertype': (user_type_udf, StringType()),
-        'gender': (gender_udf, StringType()),
-        'start_station_latitude': (angle_udf, FloatType()),
-        'start_station_longitude': (angle_udf, FloatType()),
-        'end_station_latitude': (angle_udf, FloatType()),
-        'end_station_longitude': (angle_udf, FloatType())
+        'start_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'end_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'tripduration': UserDefinedFunction(trip_duration_udf, IntegerType()),
+        'usertype': UserDefinedFunction(user_type_udf, StringType()),
+        'gender': UserDefinedFunction(gender_udf, StringType()),
+        'start_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'start_station_longitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_longitude': UserDefinedFunction(angle_udf, FloatType())
```

Comment:
What is the point of adding UserDefinedFunction on every line here instead of just once below?

Reply:
What's the point of storing the initialization parameters if its just going to be used to create a UserDefinedFunction? 

This also makes it much cleaner for the multi-parameter UDFs, specifically bc I'm storing the UDF and its parameters in a dictionary. Also this seems a bit easier to understand in the tutorial.

What do you think?